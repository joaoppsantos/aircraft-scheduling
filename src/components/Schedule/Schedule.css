.schedule {
    height: 4vh;
    width: 100%;
    background-color: lavender;
    display: flex;
}

.hour-schedule {
    height: 4vh;
    width: 4vh;
    border: 1px solid dimgrey;
    font-size: 15px;
    display: flex;
}

@media screen and (max-width: 670px) {
	.schedule {
        margin: 0 auto;
	}
}
