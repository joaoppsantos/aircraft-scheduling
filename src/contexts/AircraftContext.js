import { createContext, useEffect, useState } from 'react';
import request from '../request';

export const AircraftContext = createContext({});

export function AircraftProvider({ children }) {
	const [aircrafts, setAircrafts] = useState([]);
	const [isAircraftLoading, setIsAircraftLoading] = useState(true);
	const [currentAircraft, setCurrentAircraft] = useState(undefined);
	const [currentPercentage, setCurrentPercentage] = useState(0);

	const setPercentage = percentage => {
		setCurrentPercentage(percentage);
	};
	const selectAircraft = aircraft => {
		setCurrentAircraft(aircraft);
	};

	useEffect(() => {
		request.get(`/aircrafts?limit=5&offset=0`).then((response) => {
			setAircrafts(response.data.data);
			setIsAircraftLoading(false);
		});
	}, []);

	return (
		<AircraftContext.Provider value={{
			aircrafts,
			isAircraftLoading,
			currentAircraft,
			selectAircraft,
			currentPercentage,
			setPercentage
		}}>
			{children}
		</AircraftContext.Provider>
	)
}
