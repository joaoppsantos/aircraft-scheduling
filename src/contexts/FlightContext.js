import request from '../request';
import { createContext, useEffect, useState } from 'react';

export function FlightProvider({ children }) {
	const [flights, setFlights] = useState([]);
	const [activeFlights, setActiveFlights] = useState([]);
	const [offset, setOffset] = useState(-1);
	const [hasFlightsRemaining, setHasFlightsRemaining] = useState(true);
	const paginationLimit = 25;

	const toggleActiveFlight = (flight) => {
		let isFlightOnList = false;

		activeFlights.forEach((activeInfo) => {
			if (activeInfo.id === flight.id) {
				isFlightOnList = true;
			}
		});

		isFlightOnList ? 
			setActiveFlights(activeFlights.filter(activeInfo => activeInfo.id !== flight.id))
			: setActiveFlights([...activeFlights, flight]);
	}

	useEffect(() => {
		if (!hasFlightsRemaining) {
			return;
		}

		async function fetchData() {
			try {
				const newOffset = offset + 1;
				const response = await request.get(`/flights?limit=${paginationLimit}&offset=${newOffset}`);

				setOffset(newOffset);
				setFlights([...flights, ...response?.data?.data]);

				if (paginationLimit * (newOffset + 1) > response?.data?.pagination?.total) {
					setHasFlightsRemaining(false);
				}
			} catch (error) {
				console.log('Failed to load !');
			};
		}

		fetchData();
	// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [flights, hasFlightsRemaining]);

	return (
		<FlightContext.Provider value={{
			flights,
			toggleActiveFlight,
			activeFlights
		}}>
			{ children }
		</FlightContext.Provider>
	)
}

export const FlightContext = createContext({});