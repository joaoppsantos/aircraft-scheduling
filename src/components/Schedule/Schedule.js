import './Schedule.css';
import { useContext, useEffect } from 'react';
import { FlightContext } from '../../contexts/FlightContext';

const Schedule = () => {
    const {
        activeFlights,
    } = useContext(FlightContext);

    const arrayList = new Array(24).fill();
    const convertInHours = (seconds) => seconds/3600;

    useEffect(() => {
        let offsetWidth = document.getElementById('schedule_id')?.offsetWidth,
            offsetLeft  = document.getElementById('schedule_id')?.offsetLeft;

        const scheduleId = document.getElementById("schedule");
        const e = scheduleId.querySelectorAll("div");

        for (let i = 0; i < e.length; i++) {
            if (!!e[i]) {
                e[i].remove();
            }
        }

        activeFlights.forEach(flight => {
            let departureWidth = convertInHours(flight.departuretime),
                arrivalWidth = convertInHours(flight.arrivaltime),
                scheduledService = document.createElement("div"),
                turnaroundTime = document.createElement("div");

            let setDepartureWidth = offsetWidth / 24 * departureWidth + offsetLeft,
                setArrivalWidth = offsetWidth / 24 * arrivalWidth + offsetLeft,
                flightWidth = (setArrivalWidth - setDepartureWidth);

            scheduledService.setAttribute("id", flight.id);
            turnaroundTime.setAttribute("id", flight.id);
            scheduledService.setAttribute("style", "background-color: #6fdb95; height: 35px; margin-top: 3px; position: absolute; width: "+flightWidth+"px; left: "+setDepartureWidth+"px;");
            turnaroundTime.setAttribute("style", "background-color: #ae59e3; height: 35px; margin-top: 3px; position: absolute; width: "+15+"px; left: "+setArrivalWidth+"px;");
            document.getElementById("schedule").appendChild(scheduledService);
            document.getElementById("schedule").appendChild(turnaroundTime);
        })
	}, [activeFlights])

    return (
        <div className="schedule" id="schedule">
            {arrayList.map((_,index) => (
                <span key={index} className="hour-schedule">
                    {index}
                </span>
            ))}
        </div>
    );
}

export default Schedule;
