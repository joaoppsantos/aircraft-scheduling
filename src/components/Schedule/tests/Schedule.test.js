import * as React from 'react';
import { render } from '@testing-library/react';
import Schedule from '../Schedule';
import { FlightContext } from '../../../contexts/FlightContext';
import { flightMock } from '../../../tests/mocks/mocks';

describe('<Schedule />', () => {
    it('should render correctly', () => {
        const mockFlightContext = {
            activeFlights: [
                flightMock,
            ],
        };

        const { container } = render(
            <FlightContext.Provider value={mockFlightContext}>
                <Schedule {...flightMock} />
            </FlightContext.Provider>
        );

        expect(container).toMatchSnapshot();
    });
});
