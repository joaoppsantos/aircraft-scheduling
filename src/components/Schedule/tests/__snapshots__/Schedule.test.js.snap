// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Schedule /> should render correctly 1`] = `
<div>
  <div
    class="schedule"
    id="schedule"
  >
    <span
      class="hour-schedule"
    >
      0
    </span>
    <span
      class="hour-schedule"
    >
      1
    </span>
    <span
      class="hour-schedule"
    >
      2
    </span>
    <span
      class="hour-schedule"
    >
      3
    </span>
    <span
      class="hour-schedule"
    >
      4
    </span>
    <span
      class="hour-schedule"
    >
      5
    </span>
    <span
      class="hour-schedule"
    >
      6
    </span>
    <span
      class="hour-schedule"
    >
      7
    </span>
    <span
      class="hour-schedule"
    >
      8
    </span>
    <span
      class="hour-schedule"
    >
      9
    </span>
    <span
      class="hour-schedule"
    >
      10
    </span>
    <span
      class="hour-schedule"
    >
      11
    </span>
    <span
      class="hour-schedule"
    >
      12
    </span>
    <span
      class="hour-schedule"
    >
      13
    </span>
    <span
      class="hour-schedule"
    >
      14
    </span>
    <span
      class="hour-schedule"
    >
      15
    </span>
    <span
      class="hour-schedule"
    >
      16
    </span>
    <span
      class="hour-schedule"
    >
      17
    </span>
    <span
      class="hour-schedule"
    >
      18
    </span>
    <span
      class="hour-schedule"
    >
      19
    </span>
    <span
      class="hour-schedule"
    >
      20
    </span>
    <span
      class="hour-schedule"
    >
      21
    </span>
    <span
      class="hour-schedule"
    >
      22
    </span>
    <span
      class="hour-schedule"
    >
      23
    </span>
    <div
      id="AS1001"
      style="background-color: #6fdb95; height: 35px; margin-top: 3px; position: absolute; width: NaNpx; left: NaNpx;"
    />
    <div
      id="AS1001"
      style="background-color: #ae59e3; height: 35px; margin-top: 3px; position: absolute; width: 15px; left: NaNpx;"
    />
  </div>
</div>
`;
